File: android/src/main/kotlin/net/spross/app/DrillEffects.kt
package net.spross.app

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import net.spross.kern.session.AdvanceTier
import net.spross.kern.session.ToneKind
import net.spross.kern.trainer.DrillEffect

/**
 * The platform half both endless drills share: the beat that is armed and the four acts an
 * effect asks for.
 *
 * The state machines stay apart — a heard glyph and a typed numeral share no grammar — but
 * what kern asks of the world outside them is one list, and a second copy of it is how two
 * beats drift apart.
 */

/**
 * The armed beat before a run moves on.
 *
 * Where a screen reader is reading the screen a timed change is hostile — it truncates the
 * announcement and moves the page under the user — so no beat is ever armed there and
 * [awaitsConfirm] puts an explicit button in its place.
 */
class DrillBeat(private val screenReaderOn: () -> Boolean) {

    /** The tier waiting to elapse; null once it fired or was cancelled. */
    var tier by mutableStateOf<AdvanceTier?>(null)
        private set

    /**
     * Bumped by every arming. What a timer effect keys on: two beats in a row can be the
     * same tier, and a key that compares equal would never fire the second one.
     */
    var token by mutableStateOf(0)
        private set

    var awaitsConfirm by mutableStateOf(false)
        private set

    fun arm(next: AdvanceTier) {
        if (screenReaderOn()) {
            tier = null
            awaitsConfirm = true
            return
        }
        awaitsConfirm = false
        tier = next
        token += 1
    }

    fun cancel() {
        tier = null
        awaitsConfirm = false
    }

    /** The beat elapsed: it is spent whether or not the run had anything to book. */
    fun spend() {
        tier = null
    }
}

/**
 * What a reduction's effects become on this device. Timers, focus and playback are the
 * platform's; WHICH branch waits and which moves on is the run's rule, which is what the
 * effects say.
 */
class DrillActs(
    val beat: DrillBeat,
    /** The verdict's cue — a haptic here, since this app bundles no chimes. */
    private val onTone: (ToneKind) -> Unit,
    /** A pause that waits for a tap must give the keyboard back, or it covers the button. */
    private val onReleaseFocus: () -> Unit,
    /** Cut whatever is sounding: the reading belongs to the question being left. */
    private val onSilence: () -> Unit,
) {
    fun carryOut(effects: List<DrillEffect>) {
        for (effect in effects) {
            when (effect) {
                is DrillEffect.ArmAdvance -> beat.arm(effect.tier)
                DrillEffect.CancelAdvance -> beat.cancel()
                is DrillEffect.Tone -> onTone(effect.kind)
                DrillEffect.ReleaseFocus -> onReleaseFocus()
                DrillEffect.Silence -> onSilence()
            }
        }
    }
}

File: App/Sources/Design/DrillEffects.swift
import SprossKern

/// What a drill reduction's effects become on this device. Timers, focus and
/// playback are the platform's; WHICH branch waits and which moves on is the
/// run's rule, which is what the effects say. The three endless drills keep
/// their own state machines — a heard glyph and a typed numeral share no
/// grammar — but the acts they ask of the world outside them are one list,
/// as they are on Android (`DrillActs.carryOut`).
@MainActor
enum DrillEffects {

    /// Carry out one effect. `advance` is the caller's own armed beat, so a
    /// view that cancels or re-arms elsewhere still holds a single task.
    static func apply(_ effect: DrillEffect,
                      advance: inout Task<Void, Never>?,
                      onAdvance: @escaping @MainActor () -> Void,
                      releaseFocus: () -> Void,
                      silence: () -> Void) {
        switch onEnum(of: effect) {
        case .armAdvance(let beat):
            // why: AutoAdvance skips the timer under a screen reader — it
            // truncates the correctness announcement and moves the screen under
            // the user, and the branches render "Weiter" there instead.
            AutoAdvance.schedule(beat.tier, &advance, action: onAdvance)
        case .cancelAdvance:
            advance?.cancel()
        case .tone(let cue):
            Sound.play(cue.kind)
        case .releaseFocus:
            // why: a pause that waits for a tap must not hold the keyboard —
            // it covers the button the pause is waiting for.
            releaseFocus()
        case .silence:
            // why: the reading belongs to the question being left.
            silence()
        }
    }
}
